### PR TITLE
1. Using project SA as default. 2. Adding SA binding for default KSA in GKE cluster

### PR DESCRIPTION
**Pull request by soumyapani**

### CODE REVIEW

**Review by sdlin — APPROVED** [review]
In order for the Kubernetes Service Account to impersonate the IAM service account, the KSA will need to be annotated w/ the email address of the IAM SA.

**Comment by soumyapani** [comment]
> In order for the Kubernetes Service Account to impersonate the IAM service account, the KSA will need to be annotated w/ the email address of the IAM SA.

That is correct. Currently I am expecting annotation to be done by the user using the kubernetes yaml files. I am looking into doing that using terraform but it is not straightforward.

**Comment by sdlin** [comment]
> > In order for the Kubernetes Service Account to impersonate the IAM service account, the KSA will need to be annotated w/ the email address of the IAM SA.
> 
> That is correct. Currently I am expecting annotation to be done by the user using the kubernetes yaml files. I am looking into doing that using terraform but it is not straightforward.

How to do this using a yaml file?

I did it using this command:
```
kubectl annotate serviceaccount default --namespace default iam.gke.io/gcp-service-account=455207029971-compute@developer.gserviceaccount.com
```

In terraform, there's a kubernetes provider with a `kubernetes_service_account` resource that can also do this:
https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/service_account

**Comment by soumyapani** [comment]
> > > In order for the Kubernetes Service Account to impersonate the IAM service account, the KSA will need to be annotated w/ the email address of the IAM SA.
> > 
> > 
> > That is correct. Currently I am expecting annotation to be done by the user using the kubernetes yaml files. I am looking into doing that using terraform but it is not straightforward.
> 
> How to do this using a yaml file?
> 
> I did it using this command:
> 
> ```
> kubectl annotate serviceaccount default --namespace default iam.gke.io/gcp-service-account=455207029971-compute@developer.gserviceaccount.com
> ```
> 
> In terraform, there's a kubernetes provider with a `kubernetes_service_account` resource that can also do this: https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/service_account

you can have something like this in the yaml file
````yaml
# Definition of the namespace for the cluster, and related configurations.
# The file creates:
# - namespace
# - ServiceAccount binding with the namespace
#

apiVersion: v1
kind: Namespace
metadata:
  name: default

---
# Define Kubernetes Service Accounts to use with workload identity on the namespace.
# This is the same as using kubectl annotate serviceaccount
apiVersion: v1
kind: ServiceAccount
metadata:
  name: default
  namespace: default
  annotations:
    # Explicitly mapping the Kubernetes Service account to GSA.
    iam.gke.io/gcp-service-account: ${GSA-name}
````

The problem with `kubernetes_service_account` is it cannot be used with our current module in a single `terraform apply` session due to terraform limitations (happy to discuss more on this offline). However we can have it as a separate terraform config (`main.tf` file) and users have to do `terraform apply` on that separately to annotate the account. At this point, we are actually not creating GKE resources, we are talking to the GKE cluster directly and simulating `kubectl` calls.  So I am not sure if we should include it in the provisioning tool. Users would also want to create their own KSA annotate it differently.

**Review by stevenBorisko — APPROVED** [review]

**Comment by soumyapani** [comment]
> > > > In order for the Kubernetes Service Account to impersonate the IAM service account, the KSA will need to be annotated w/ the email address of the IAM SA.
> > > 
> > > 
> > > That is correct. Currently I am expecting annotation to be done by the user using the kubernetes yaml files. I am looking into doing that using terraform but it is not straightforward.
> > 
> > 
> > How to do this using a yaml file?
> > I did it using this command:
> > ```
> > kubectl annotate serviceaccount default --namespace default iam.gke.io/gcp-service-account=455207029971-compute@developer.gserviceaccount.com
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In terraform, there's a kubernetes provider with a `kubernetes_service_account` resource that can also do this: https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/service_account
> 
> you can have something like this in the yaml file
> 
> ```yaml
> # Definition of the namespace for the cluster, and related configurations.
> # The file creates:
> # - namespace
> # - ServiceAccount binding with the namespace
> #
> 
> apiVersion: v1
> kind: Namespace
> metadata:
>   name: default
> 
> ---
> # Define Kubernetes Service Accounts to use with workload identity on the namespace.
> # This is the same as using kubectl annotate serviceaccount
> apiVersion: v1
> kind: ServiceAccount
> metadata:
>   name: default
>   namespace: default
>   annotations:
>     # Explicitly mapping the Kubernetes Service account to GSA.
>     iam.gke.io/gcp-service-account: ${GSA-name}
> ```
> 
> The problem with `kubernetes_service_account` is it cannot be used with our current module in a single `terraform apply` session due to terraform limitations (happy to discuss more on this offline). However we can have it as a separate terraform config (`main.tf` file) and users have to do `terraform apply` on that separately to annotate the account. At this point, we are actually not creating GKE resources, we are talking to the GKE cluster directly and simulating `kubectl` calls. So I am not sure if we should include it in the provisioning tool. Users would also want to create their own KSA annotate it differently.

I have added the support to do both GSA account binding and KSA annotation for GKE cluster after it is created.

File: README.md [hunk 1]
```diff
@@ -100,7 +100,28 @@ The optional parameters are:
...
+   ```tf
+    enable_k8s_setup                     = bool
+    kubernetes_service_account_name      = string
+    kubernetes_service_account_namespace = string
+    node_service_account                 = string
+   ```
+   * `enable_k8s_setup`: This is the flag to enable kubernetes setup. The default value is
+         true if the `ORCHESTRATOR_TYPE` is `gke`.
+   * `kubernetes_service_account_name`: The kubernetes service account (KSA) name. The
+         value is `aiinfra-gke-sa`.
```
Comment (stevenBorisko) [inline README.md:116]: Possible small typo?:
```diff
- The value is `aiinfra-gke-sa`.
+ The default value is `aiinfra-gke-sa`.
```

File: aiinfra-cluster/main.tf [hunk 2]
```diff
@@ -194,3 +201,17 @@ module "aiinfra-default-dashboard" {
...
+  gke_conn            = {
+    gke_cluster_endpoint           = module.aiinfra-compute.gke_cluster_endpoint
+    gke_certificate_authority_data = module.aiinfra-compute.gke_certificate_authority_data
+    gke_token                      = module.aiinfra-compute.gke_token
+  }
+  enable_k8s_setup                     = local.kubernetes_setup_config.enable_k8s_setup
+  kubernetes_service_account_name      = local.kubernetes_setup_config.kubernetes_service_account_name
+  kubernetes_service_account_namespace = local.kubernetes_setup_config.kubernetes_service_account_namespace
+  node_service_account = local.kubernetes_setup_config.node_service_account
+}
```
Comment (stevenBorisko) [inline aiinfra-cluster/main.tf:217]: does this need a `count = var.orchestrator_type == "gke" ? 1 : 0` in it or will the `provider "kubernetes"` just not not worry about it?

File: aiinfra-cluster/outputs.tf [hunk 3]
```diff
@@ -18,3 +18,18 @@ output "dashboard_instructions" {
...
+
+output "gke_cluster_connection" {
+  description = "The GKE cluster connection Information"
+  value = var.orchestrator_type == "gke" ? {
+    gke_cluster_endpoint           = module.aiinfra-compute.gke_cluster_endpoint
+    gke_certificate_authority_data = module.aiinfra-compute.gke_certificate_authority_data
+    gke_token                      = module.aiinfra-compute.gke_token
+  } : null
+  sensitive = true
+}
```
Comment (stevenBorisko) [inline aiinfra-cluster/outputs.tf:35]: Does `sensitive = true` make it never show up in the terminal? If so, is there a reason to have it in the root module's outputs?

**Review by stevenBorisko — APPROVED** [review]
Great workaround. I had no idea you could put a provider in a submodule. I approved just in case I don't know what I'm talking about with the comments I added

**Pull request merged**